frontendGraph/src/main/java/com/virtuslab/gitmachete/frontend/graph/print/PrintElementGeneratorImpl.java

```diff
@@ -0,0 +1,107 @@
+    assert rowIndex >= 0;
+    PrintElementBuilder builder = new PrintElementBuilder(rowIndex);
+    collectElements(rowIndex, builder);
+    return builder.build();
+  }
+
+  private void collectElements(@NonNegative int rowIndex, PrintElementBuilder builder) {
+    GraphNode graphNode = new GraphNode(rowIndex);
+    IGraphNode graphElement = repositoryGraph.getGraphNode(rowIndex);
+    int position = graphElement.getIndentLevel();
+
+    repositoryGraph.getVisibleEdgesWithPositions(rowIndex).forEach(edgeAndPos -> {
+      builder.consumeUpEdge(edgeAndPos._1(), edgeAndPos._2());
+      builder.consumeDownEdge(edgeAndPos._1(), edgeAndPos._2());
+    });
+
+    java.util.List<GraphEdge> adjacentEdges = repositoryGraph.getAdjacentEdges(rowIndex, EdgeFilter.ALL);
+    for (GraphEdge edge : adjacentEdges) {
+      Integer downNodeIndex = edge.getDownNodeIndex();
+      Integer upNodeIndex = edge.getUpNodeIndex();
+      if (downNodeIndex != null && downNodeIndex == rowIndex) {
+        builder.consumeUpEdge(edge, position);
+      }
+      if (upNodeIndex != null && upNodeIndex == rowIndex) {
+        builder.consumeDownEdge(edge, position);
+      }
+    }
+
+    int nodeAndItsDownEdgePos = position;
+    if (graphElement.isBranch() && !((BranchNode) graphElement).getBranch().isRootBranch()) {
+      builder.consumeRightEdge(new GraphEdge(rowIndex, rowIndex, /* targetId */ null, GraphEdgeType.USUAL), position);
+      nodeAndItsDownEdgePos++;
+    }
+
+    builder.consumeNode(graphNode, nodeAndItsDownEdgePos);
+    if (graphElement.hasChildNode()) {
+      builder.consumeDownEdge(new GraphEdge(rowIndex, rowIndex + 1, /* targetId */ null, GraphEdgeType.USUAL),
+          nodeAndItsDownEdgePos);
+    }
+  }
+
+  @RequiredArgsConstructor
+  private final class PrintElementBuilder {
+    private final ArrayList<PrintElementWithGraphElement> result = new ArrayList<>();
+    private final ArrayList<PrintElementWithGraphElement> nodes = new ArrayList<>();
+    @NonNegative
+    private final int rowIndex;
+
+    public void consumeNode(GraphNode node, @NonNegative int position) {
+      nodes.add(new SimplePrintElementImpl(rowIndex, position, node, printElementManager));
+    }
+
+    public void consumeDownEdge(GraphEdge edge, @NonNegative int position) {
+      result.add(
+          new EdgePrintElementImpl(rowIndex, position, position, Type.DOWN, edge, /* hasArrow */ false,
+              printElementManager));
+    }
+
+    public void consumeUpEdge(GraphEdge edge, @NonNegative int position) {
+      result.add(
+          new EdgePrintElementImpl(rowIndex, position, position, Type.UP, edge, /* hasArrow */ false,
+              printElementManager));
+    }
+
+    public void consumeRightEdge(GraphEdge edge, @NonNegative int position) {
+      result.add(new RightEdgePrintElement(rowIndex, position, edge, printElementManager));
+    }
+
+    public Collection<PrintElementWithGraphElement> build() {
+      result.addAll(nodes);
```

Comment:
Hmm why then not add to `result` directly from the very beginning? :thinking:

Reply:
Although I see that's pretty unlikely given that the builder class is private & the builder instance is quickly discarded... up to you then